pkg/agents/team4/finalAgent/treaty.go

```diff
@@ -0,0 +1,210 @@
+		fallthrough
+	case treaty.Condition() == messages.AvailableFood && treaty.Request() == messages.LeavePercentFood && int(float64(100-treaty.RequestValue()/100))*treaty.ConditionValue() <= a.HealthInfo().HPReqCToW*2:
+		fallthrough
+	case treaty.Condition() == messages.AvailableFood && (treaty.ConditionOp() == messages.LT || treaty.ConditionOp() == messages.LE):
+		fallthrough
+	case treaty.Condition() == messages.Floor && (treaty.ConditionOp() == messages.GT || treaty.ConditionOp() == messages.GE):
+		fallthrough
+	case a.HP() < a.HealthInfo().HPCritical: // Reject all treaties at that point in time when we're below critical health (all treaty requests are relevant to food only so this applies).
+		fallthrough
+	case treaty.Condition() == messages.AvailableFood && treaty.Request() == messages.LeaveAmountFood && treaty.ConditionValue()-a.params.foodToEat["selfless"][2] <= treaty.RequestValue(): //the worst we want to be is in selfless
+		fallthrough
+	case treaty.Condition() == messages.AvailableFood && treaty.Request() == messages.LeavePercentFood && treaty.ConditionValue()-a.params.foodToEat["selfless"][2] <= treaty.ConditionValue()*treaty.RequestValue()/100: //the worst we want to be is in selfless
+		fallthrough
+	case treaty.Duration() >= a.HealthInfo().MaxDayCritical:
+		fallthrough
+	case int(a.params.globalTrust) < a.params.globalTrustLimits[0]:
+		a.rejectTreaty(msg)
+		return
+	}
+
+	// If it passes our conditions above then we accept the treaty
+	treaty.SignTreaty()
+	a.AddTreaty(treaty)
+	reply := msg.Reply(a.ID(), a.Floor(), msg.SenderFloor(), true)
+	a.SendMessage(reply)
+	a.Log("Accepted treaty", infra.Fields{"proposerID": msg.SenderID(), "proposerFloor": msg.SenderFloor(), "treatyID": msg.TreatyID()})
+	// We propogate the treaty to the floor above/below if we have signed the treaty
+	a.propogateTreaty(msg)
+}
+
+/*------------------------REJECT A TREATY------------------------*/
+
+func (a *CustomAgentEvo) rejectTreaty(msg messages.ProposeTreatyMessage) {
+	reply := msg.Reply(a.ID(), a.Floor(), msg.SenderFloor(), false)
+	a.SendMessage(reply)
+	a.Log("Rejected treaty", infra.Fields{"proposerID": msg.SenderID(), "proposerFloor": msg.SenderFloor(), "treatyID": msg.TreatyID()})
+}
+
+/*------------------------PROPOGATE A TREATY TO THE FLOOR ABOVE AND BELOW------------------------*/
+
+func (a *CustomAgentEvo) propogateTreaty(msg messages.ProposeTreatyMessage) {
+	treaty := msg.Treaty()
+	toFloor := a.Floor() + 1
+	if msg.SenderFloor() > a.Floor() {
+		toFloor = a.Floor() - 1
+	}
+	propogateTreaty := messages.NewProposalMessage(a.ID(), a.Floor(), toFloor, treaty)
+	a.SendMessage(propogateTreaty)
+}
+
+/*------------------------HANDLING OTHER AGENTS REPSONSES TO OUR TREATY PROPOSALS------------------------*/
+
+func (a *CustomAgentEvo) HandleTreatyResponse(msg messages.TreatyResponseMessage) {
+	if msg.Response() { // Check if there is a response
```

Comment:
The comment is wrong - `msg.Response()` indicates whether or not they signed the treaty?